nixos/modules/services/web-apps/invidious.nix

```diff
@@ -7,17 +7,31 @@ let
 
   settingsFile = settingsFormat.generate "invidious-settings" cfg.settings;
 
+  finalHmacKeyFile =
+    if cfg.hmacKeyFile != null
+    then cfg.hmacKeyFile
+    else "/var/lib/invidious/hmac_key";
+  generateHmac = cfg.hmacKeyFile == null;
+
   serviceConfig = {
     systemd.services.invidious = {
       description = "Invidious (An alternative YouTube front-end)";
       wants = [ "network-online.target" ];
       after = [ "network-online.target" ];
       wantedBy = [ "multi-user.target" ];
 
+      preStart = lib.optionalString generateHmac ''
+        if [[ ! -r "${finalHmacKeyFile}" ]]; then
+          ${pkgs.pwgen}/bin/pwgen 20 1 > "${finalHmacKeyFile}"
+          chmod 0600 "${finalHmacKeyFile}"
+        fi
+      '';
+
       script =
         let
           jqFilter = "."
-            + lib.optionalString (cfg.database.host != null) "[0].db.password = \"'\"'\"$(cat ${lib.escapeShellArg cfg.database.passwordFile})\"'\"'\""
+            + " | .[0].hmac_key = \"'\"'\"$(cat ${lib.escapeShellArg finalHmacKeyFile})\"'\"'\""
+            + lib.optionalString (cfg.database.host != null) " | .[0].db.password = \"'\"'\"$(cat ${lib.escapeShellArg cfg.database.passwordFile})\"'\"'\""
             + " | .[0]"
             + lib.optionalString (cfg.extraSettingsFile != null) " * .[1]";
           jqFiles = [ settingsFile ] ++ lib.optional (cfg.extraSettingsFile != null) cfg.extraSettingsFile;
```

Comment:
This is a real mess, if you have the time I'd really appreciate a refactoring of this. Ideally avoiding any escaping madness by having a constant `jq` expression and using `jq`'s `--arg` combined with `lib.escapeShellArg` to pass in values.